ringpop.thrift-gen

```diff
@@ -0,0 +1,170 @@
+{{ define "arguments" }}ctx {{ contextType }}{{ range .Arguments }}, {{ .Name }} {{ .Type | goType }}{{ end }}{{ end}}
+{{ define "call" }}ctx {{ range .Arguments }}, {{ .Name }}{{ end }}{{ end}}
+{{ define "returnType" }}{{ if .ReturnType }}r {{ .ReturnType | goType }}, {{ end }}err error{{ end}}
+
+{{ $Root := . }}
+package {{ .Package }}
+
+import (
+    "errors"
+    "fmt"
+    "sync"
+
+    "github.com/uber/ringpop-go"
+    "github.com/uber/tchannel-go"
+    "github.com/uber/tchannel-go/thrift"
+)
+
+{{ range .AST.Services }}
+{{ $Service := . }}
+{{ $TChanInterface :=  print "TChan" $Service.Name }}
+{{ $Adapter :=  print "Ringpop" $Service.Name "Adapter" }}
+{{ $ConfigName :=  print $Service.Name "Configuration" }}
+
+// ------ Generated code for {{ $Service.Name }} ------
+type {{ $Adapter }} struct {
+    impl        {{ $TChanInterface }}
+    ringpop     ringpop.RingpopInterface
+    ch          *tchannel.Channel
+    config      {{ $ConfigName }}
+    router      ringpop.Router
+}
+
+type {{ $ConfigName }} struct {
+    // structs for all service endpoints containing forwarding behavior
+    {{ range .Methods }}
+        {{ .Name | goPublicName }} *{{ $Service.Name }}{{ .Name | goPublicName }}Configuration
+    {{ end }}
+}
+// TODO make validation function on the config
+
+type tchan{{ $Service.Name }}Adaptable interface {
+    SetAdapter({{ $TChanInterface }})
+}
+
+func New{{ $Adapter }}(
+    impl   {{ $TChanInterface }},
+    rp     ringpop.RingpopInterface,
+    ch     *tchannel.Channel,
+    config {{ $ConfigName }},
+) *{{ $Adapter }} {
+    adapter := &{{ $Adapter }}{
+        impl:    impl,
+        ringpop: rp,
+        ch:      ch,
+        config:  config,
+    }
+    // create ringpop router for routing based on ring membership
+    adapter.router = ringpop.NewRouter(rp, adapter)
+
+    // notify implementation it is adapted
+    if adaptable, ok := impl.(tchan{{ $Service.Name }}Adaptable); ok {
```

Comment:
That's a good point, since the user has a reference to the adapter, they could call `SetAdapter` themselves. Putting this in the generated code means that if the user had a Thrift method "SetAdapter", they couldn't use this.